### ISSUE TITLE
Paramaters matched to JSON fields are the wrong type

### ISSUE THREAD

**Issue by kyleconroy** [sqlc-dev/sqlc#743]
These parameters should end up as `interface{}`, not `json.RawMessage`.

```sql
CREATE TABLE "user" (
    "id" INT PRIMARY KEY GENERATED ALWAYS AS IDENTITY,
    "metadata" JSONB
)

-- name: FindByAddress :one
SELECT * FROM "user" WHERE "metadata"->>'address1' = $1 LIMIT 1
```

```go
type FindByAddressParams struct {
	metadata json.RawMessage
}
```

- [Initial report](https://github.com/kyleconroy/sqlc/discussions/738?converting=1)

https://play.sqlc.dev/p/cd6227106fb8fb969d1fc9699dd9c03046e26578c589268e30d23573f67037bb

**Comment by Abdillah-Epi** [comment]
Hi @kyleconroy, I have the same problem as in issue #738, I have table like this:

```sql
CREATE TABLE "user" (
    "id" INT PRIMARY KEY GENERATED ALWAYS AS IDENTITY,
    "metadata" JSONB
)
```

but instead of having an object, I have an array of users:

```text
[
    {
        "lic":"1137868E",
        "firstname":"Joe",
        "lastname":"Bag"
    },
    {
        "lic":"1295364R",
        "firstname":"HA",
        "lastname":"Dégue"
    },
]
```

then I wrote a query to get a user by "lic"

```sql
-- name: FindUserByLic :one
SELECT * FROM user AS u WHERE u.deleted_at IS NULL AND u.metadata @> $1; -- > $1 suppose to be like this '[{"lic":"1295364R"}]'
```

and I got this parameter:

```go
type FindUserByLic struct {
	metadata json.RawMessage
}
```

here is my code:
```go
	lic, err := json.Marshal(fmt.Sprintf(`'[{"lic":"%s"}]'`, req.Lic))
	if err != nil {
		return err
	}
	arg := db.FindUserByLicParams{
		Metadata: lic,
	}

	contests, err := server.store.FindUserByLic(ctx, arg) //find user by lic
	if err != nil {
		return err
	}
```

but in the end I had an empty array :/

**Comment by speix** [comment]
That is not a bug, I have a solution for that or any parametrized representation of jsonb operations.

First you need to define the datatype of your parameter on your sqlc definitions, for example:

`... and cl.records @> sqlc.arg(records)::jsonb -- Here you expect to input: '[{"id":"1234567890"}]'`

Then in your go code, you would have to have an interface conversion to JSONB:
```
func InterfaceToJSONB(data interface{}) pgtype.JSONB {
	var jsonB pgtype.JSONB
	err := jsonB.Set(data)
	return jsonB
}
```

Last, running your query, you have this:
```
param3 := "1234567890"
someData, err := q.SomeQueryName(ctx, db.SomeQueryParamsParams{
		ParamName1:  param1,
		ParamName2: param2,
		Listings: InterfaceToJSONB(`[{"id":"` + param3 + `"}]`),
	})
```

**Comment by kyleconroy** [comment]
This is fixed in [v1.23.0](https://docs.sqlc.dev/en/latest/reference/changelog.html#v1-23-0) by enabling the [database-backed query analyzer](https://docs.sqlc.dev/en/latest/howto/generate.html#enhanced-analysis-with-managed-databases). We added a [test case](https://github.com/sqlc-dev/sqlc/tree/main/internal/endtoend/testdata/json_param_type/postgresql/pgx) for this issue so it won’t break in the future.

You can play around with the [working example on the playground](https://play.sqlc.dev/p/0ffa93167e503ed8f4a55e764bd5a6cd3f2b61b8cc0a3d4f32ec3a7b9cd21465)
